keps/sig-node/3953-dynamic-node-resize/README.md

```diff
@@ -0,0 +1,776 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The proposal aims at enabling dynamic node resizing. This will help in updating cluster resource capacity by just resizing compute resources of nodes rather than adding new node or removing existing node from a cluster.
+The updated node configurations are to be reflected at the node and cluster levels automatically without the need to reset the kubelet.
+
+This proposal also aims to improve the initialization and reinitialization of resource managers, such as the CPU manager and memory manager, in response to changes in a node's CPU and memory configurations.
+
+## Motivation
+In a typical Kubernetes environment, the cluster resources may need to be altered due to following reasons:
+- Incorrect resource assignment during cluster creation.
+- Increased workload over time, leading to the need for additional resources in the cluster.
+- Decreased workload over time, leading to resource underutilization in the cluster.
+
+To handle these scenarios, we can:
+- Horizontally scale up or down the cluster by adding or removing compute nodes.
+- Vertically scale up or down the cluster by increasing or decreasing node capacity. However, currently, the workaround for capturing node resizing in the cluster involves restarting the Kubelet.
+
+Dynamic node resizing will provide advantages in scenarios such as:
+- Handling resource demand with a limited set of nodes by increasing the capacity of existing nodes instead of creating new nodes.
+- Creating or deleting new nodes takes more time compared to increasing or decreasing the capacity of existing nodes.
+
+### Goals
+
+* Dynamically resize the node without restarting the kubelet.
+* Ability to reinitialize resource managers (CPU manager, memory manager) to adopt changes in node's resource.
+
+### Non-Goals
+
+* Update the autoscaler to utilize dynamic node resize.
+
+## Proposal
+
+This KEP adds a polling mechanism in kubelet to fetch the machine-information using cAdvisor, The information will be fetched periodically based on a configured time interval, after which the node status updater is responsible for updating this information at node level in the cluster.
+
+Additionally, this KEP aims to improve the initialization and reinitialization of resource managers, such as the memory manager and CPU manager, so that they can adapt to changes in node's configurations.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin, I must be able to increase the cluster resource capacity without adding a new node to the cluster.
+
+#### Story 2
+
+As a cluster admin, I must be able to decrease the cluster resource capacity without removing an existing node from the cluster.
```

Comment:
Decreasing resources may cause pod evictions.
- if a pod is a static pod, will it be stopped? If not, the allocatable resource may be under 0.

Reply:
Currently we thought of utilizing the current method https://github.com/kubernetes/kubernetes/blob/61d455ed1173cd89a98442adf4623a29c5681c58/pkg/kubelet/kubelet.go#L2519 to handle the pod admission, Reference is available here https://github.com/kubernetes/kubernetes/pull/115755